vcd/resource_vcd_rde_type.go

```diff
@@ -0,0 +1,337 @@
+	areEqual, err := areMarshaledJsonEqual([]byte(oldValue), []byte(newValue))
+	if err != nil {
+		util.Logger.Printf("[ERROR] Could not compare JSONs for computing difference of %s: %s", key, err)
+		return false
+	}
+	return areEqual
+}
+
+func resourceVcdRdeTypeCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	jsonSchema, err := getRdeTypeSchema(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	executeRdeTypeFunctionWithMutex(func() {
+		_, err = vcdClient.VCDClient.CreateRdeType(&types.DefinedEntityType{
+			Name:             d.Get("name").(string),
+			Nss:              d.Get("nss").(string),
+			Version:          d.Get("version").(string),
+			Description:      d.Get("description").(string),
+			ExternalId:       d.Get("external_id").(string),
+			InheritedVersion: d.Get("inherited_version").(string),
+			Interfaces:       convertSchemaSetToSliceOfStrings(d.Get("interface_ids").(*schema.Set)),
+			IsReadOnly:       d.Get("readonly").(bool),
+			Schema:           jsonSchema,
+			Vendor:           d.Get("vendor").(string),
+		})
+	})
+
+	if err != nil {
+		return diag.Errorf("could not create the Runtime Defined Entity Type: %s", err)
+	}
+	return resourceVcdRdeTypeRead(ctx, d, meta)
+}
+
+// getRdeTypeSchema gets the schema as string from the Terraform configuration
+func getRdeTypeSchema(d *schema.ResourceData) (map[string]interface{}, error) {
+	var jsonSchema string
+	var err error
+	if url, isUrlSet := d.GetOk("schema_url"); isUrlSet {
+		jsonSchema, err = fileFromUrlToString(url.(string), ".json")
+		if err != nil {
+			return nil, fmt.Errorf("could not download JSON schema from url %s: %s", url, err)
+		}
+	} else {
+		jsonSchema = d.Get("schema").(string)
+	}
+
+	var unmarshalledJson map[string]interface{}
+	err = json.Unmarshal([]byte(jsonSchema), &unmarshalledJson)
+	if err != nil {
+		return nil, err
+	}
+
+	return unmarshalledJson, err
+}
+
+// executeRdeTypeFunctionWithMutex executes the given function related to RDE writing with a mutex, this is because
+// RDE Type write operations suffer from race conditions (at least in API v37.2), hence more than 1 RDE Type cannot be
+// written in parallel.
+// We force to do it sequentially with a mutex.
+func executeRdeTypeFunctionWithMutex(rdeWriteFunction func()) {
+	key := "vcd_rde_type"
+	vcdMutexKV.kvLock(key)
+	rdeWriteFunction()
+	vcdMutexKV.kvUnlock(key)
+}
+
+// fileFromUrlToString checks that the given url is correct and points to a given file type,
+// if so it downloads its contents and returns it as string.
+func fileFromUrlToString(url, fileType string) (string, error) {
+	if !strings.HasSuffix(url, fileType) {
+		return "", fmt.Errorf("it was expecting the URL to point to a %s file but it was %s", fileType, url)
+	}
+	buf := new(strings.Builder)
+
+	// #nosec G107 -- The URL needs to come from a variable for this purpose
+	resp, err := http.Get(url)
```

Comment:
It is sort of stupid here, but you need to at least initialize client with timeout as it might hang forever https://medium.com/@nate510/don-t-use-go-s-default-http-client-4804cb19f779

Or maybe access the same initialized HTTP client from  go-vcloud-director `vcdClient.Client.Http.Get`

Reply:
FYI: @adezxc (just for reference)